File: FindifPathExistsinGraph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        if n == 1:
            return True
        
        # create a dictionary where the value is a list
        graph = defaultdict(list) # looks like this: {a: [b, c], b: [a]}
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)        
        
        visited = set()
        visited.add(source)
        stack = [source]

        while stack: # Iterating the stack will perform DFS
            n = stack.pop()
            if n == destination:
                return True
            for node in graph[n]:
                if node not in visited:
                    visited.add(node)
                    stack.append(node) 
        return False

        # DFS using stack
        # Time: O(n + E)
        # Space: O(n + E)
        # nodes(n) from stack and Edges(E) from the graph (ie. deafaultdict(list))

File: FindifPathExistsinGraph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        if (edges.length == 1)
            return true;
        
        Map<Integer, List<Integer>> edge = new HashMap<>();
        for(int i = 0; i < edges.length; i ++){
            int a = edges[i][0];
            int b = edges[i][1];
            edge.computeIfAbsent(a, k -> new ArrayList<>()).add(b); // make the array list if the given key does not exist
            edge.computeIfAbsent(b, k -> new ArrayList<>()).add(a);
        }

        Set<Integer> visited = new HashSet<>();
        visited.add(source);
        Stack<Integer> s = new Stack<>();
        s.push(source);

        while (s.size() > 0){
            int node = s.pop();
            if (node == destination){
                return true;
            }
            for (int neiNodes: edge.get(node)){
                if (!visited.contains(neiNodes)){
                    visited.add(neiNodes);
                    s.push(neiNodes);
                }
            }
        }
        return false;
    }

    // Time: O(N + Edges)
    // Space: O(N + Edges)
    // N number of nodes, Edges number of edges
}
